### ISSUE TITLE
Cannot read property 'length' of undefined

### ISSUE THREAD

**Issue by digaus** [jhildenbiddle/css-vars-ponyfill#1]
Trying to use it but only getting this error. Caused by: `return r.declarations.length ;`

Fixed it with
```
 if(r.declarations) {
     return r.declarations.length > 0 ;
} else {
     return false
}
```

**Comment by jhildenbiddle** [comment]
Happy to help @digaus, but I'll need more details on how you encountered this error.

- Are you including css-vars-ponyfill in a bundle via webpack/rollup/parcel/etc?
- Are you loading css-vars-ponyfill from a CDN?
- How are you calling the ponyfill?
- Are you processing CSS in `<style>` or `<link>` tags?
- If you are processing external stylesheets via `<link>` tag or `@import` rules, are they on the same domain?

This info will give me a better chance of resolving your issue.

In the meantime, I've updated the [docs](https://github.com/jhildenbiddle/css-vars-ponyfill) and added a [demo](https://codepen.io/jhildenbiddle/pen/ZxYJrR/) that you may find useful.

**Comment by halostatue** [comment]
I am also seeing this on IE11:

```
Unable to get property 'length' of undefined or null reference
 [object HTMLLinkElement]
   "Unable to get property 'length' of undefined or null reference
"
   <link href="/css/employee-app.css" rel="stylesheet"></link>
```

**Comment by jhildenbiddle** [comment]
@halostatue - Is there any way you can share the HTML and CSS code so that I can try to reproduce the problem? The page HTML and `employee-app.css` should be sufficient.

Thanks, and hopefully I will be able to resolve this quickly for you.

**Comment by halostatue** [comment]
Unfortunately, no, I can’t, not at the moment. I’m adding some debugging to try to understand what’s going on, but I cannot seem to get `yarn run build` to work inside my node_modules so I can test this.

**Comment by jhildenbiddle** [comment]
@halostatue Gotchya. If you can provide the CSS that triggers the error, I'm sure I can track down the issue fairly quickly.

In the meantime, try setting the `onlyVars` option to `false` (the default value is `true`). Details on this option can be found [here](https://www.npmjs.com/package/css-vars-ponyfill#optionsonlyvars).

**Comment by halostatue** [comment]
```
{
   [functions]: ,
   __proto__: { },
   comment: " line 57, components/common/Card.vue ",
   type: "comment"
}
```

This is debugged from `transform-css.js` where you do `rule.rules = filterVars(rule.rules).filter(r => r.declarations.length);`. It appears to be a debug-only CSS transform performed, but as stated in the opening issue, it would probably be best to do `r => r.declarations && r.declarations.length`. Testing that now.

**Comment by halostatue** [comment]
Fixed with the proposed change. I did not try the `onlyVars` setting, but this would probably fix other bits. I cannot push this live anywhere.

**Comment by jhildenbiddle** [comment]
Very much appreciate the PR @halostatue, but I would like to understand the scenario that causes the issue and account for it in the automated tests before making changes on my own or accepting PRs.

Now that you are able to build, can you share the contents of the CSS file that triggered the issue?

Thanks!

**Comment by jhildenbiddle** [comment]
@digaus Would you also be able to share the CSS that triggered the issue? As I mentioned with @halostatue , I want to understand the issue so that I can fix the bug and account for it in the automated tests. The best way to do that is testing with the CSS that is known to cause the issue in IE11.

Thanks in advance. Very much appreciate the help in tracking this down. :)

**Comment by halostatue** [comment]
I can, but it’s fairly large, and if I were to push this up publicly, the comment nodes do not appear.

So you should be able to verify this with a simple file that looks like:

```css
/* line 36, components/customer/CustomerApp.vue */
:root {
  --brand-primary: #0A4990;
  --bg-primary: #FBFBFB;
  --bg-secondary: white;
  --border-primary: #D3D3D3;
  --warning: #FF0000;
  --font-color-primary: #444444;}

/* line 360, components/common/EditBusinessProfile.vue */
.go-to {
  color: var(--brand-primary);
  text-decoration: underline;
  cursor: pointer; }
```

The issue was specifically on the `comment` nodes of the parsed CSS in development mode, and would not be present in production mode.

**Comment by halostatue** [comment]
The total file is ~3klines and is a bit of a mess (I had to edit out some scoped CSS above we have because we are doing Vue.js).

**Comment by jhildenbiddle** [comment]
Thanks for this, @halostatue .

The CSS you provided isn't responsible for the error in IE11. I've put together a quick CodePen to demonstrate:

- [CodePen](https://codepen.io/jhildenbiddle/pen/qormWE)
- [CodePen "Debug"](https://s.codepen.io/jhildenbiddle/debug/qormWE/gaAeYJoYvNdA)

I'm wondering if the scoped CSS you referred to is to blame? If you can share the file as-is, it will be easier to track down the CSS that is causing the issue. If you don't want to post it to GH, feel free to send it to me directly (firstname @ lastname dot com).

By the way, I understand that the fix that you provided should be in place *regardless* of the details of this issue. My goal is to identify the CSS that caused the `filterVars()` function to return rule(s) without a `declarations` array (which is the cause of the error). By adding that CSS to the automated tests and ensuring everything passes as expected, I'll feel more confident that we've addressed the issue completely.

Thanks again!

**Comment by halostatue** [comment]
[employee-app.css.zip](https://github.com/jhildenbiddle/css-vars-ponyfill/files/1831488/employee-app.css.zip)

I am initializing with `cssVars({ exclude: '[href*=semantic]'})` because we have an external load for Semantic UI CSS in the HTML, but aside from that, the only `link` is to the CSS file in question.

**Comment by halostatue** [comment]
I tried adding the scope (as `.go-to[data-xyz] {…}`) and things worked just fine with your example in Safari.

**Comment by jhildenbiddle** [comment]
This is exactly what I needed. :)

Comments as direct descendants of at-rules exposed the bug.

```css
@media {
    /* Comment*/
}
```

Automated tests were in place to ensure `css-vars-ponyfill` handled comments properly, but I had not accounted for this scenario. The previous tests verified that this CSS:

```css
/* Comment */
:root {
    /* Comment */
    --color: /* Comment */ red /* Comment */; /* Comment */
    /* Comment */
}
/* Comment */
p {
    /* Comment */
    color: /* Comment */ var(--color) /* Comment */; /* Comment */
    /* Comment */
}
/* Comment */
```

Produces the following output:

```css
p{color:red;}
```

I have updated the automated tests accordingly, addressed the bug in the source, and will publish an update shortly.

Thanks for taking the time to track this down. Very much appreciated!

**Comment by halostatue** [comment]
Not a problem. I look forward to the published update as it will fix (some) of the issues we are seeing with IE 11 that appear to be (mostly) CSS related. I’ll have to see what other features are failing and find polyfills for those, too.

**Comment by jhildenbiddle** [comment]
I have pushed a new version (1.2.0) to NPM with the bug fixed.

Also, apologies to @halostatue for implementing the same fix on my end without merging #2 . I made additional updates that needed to accompany the bug fix, so I just took care of it in one swoop.

- Updated automated tests to verify bug identified on this thread is fixed
- Updated CHANGELOG
- Updated Tracis-CI config to run local tests using headless Chrome instead of remote tests on SauceLabs (necessary due to Travis-CI blocking of encrypted environment variables for all repos other than main repo).

Thanks again for the time and effort!
